mesa/examples/advanced/wolf_sheep/model.py

```diff
@@ -0,0 +1,134 @@
+"""
+Wolf-Sheep Predation Model
+================================
+
+Replication of the model found in NetLogo:
+    Wilensky, U. (1997). NetLogo Wolf Sheep Predation model.
+    http://ccl.northwestern.edu/netlogo/models/WolfSheepPredation.
+    Center for Connected Learning and Computer-Based Modeling,
+    Northwestern University, Evanston, IL.
+"""
+
+import mesa
+from mesa.experimental.cell_space import OrthogonalMooreGrid
+
+from mesa.examples.advanced.wolf_sheep.agents import GrassPatch, Sheep, Wolf
+
+
+class WolfSheep(mesa.Model):
+    """
+    Wolf-Sheep Predation Model
+    """
+
+    height = 20
+    width = 20
+
+    initial_sheep = 100
+    initial_wolves = 50
+
+    sheep_reproduce = 0.04
+    wolf_reproduce = 0.05
+
+    wolf_gain_from_food = 20
+
+    grass = False
+    grass_regrowth_time = 30
+    sheep_gain_from_food = 4
+
+    description = (
+        "A model for simulating wolf and sheep (predator-prey) ecosystem modelling."
+    )
+
+    def __init__(
+        self,
+        width=20,
+        height=20,
+        initial_sheep=100,
+        initial_wolves=50,
+        sheep_reproduce=0.04,
+        wolf_reproduce=0.05,
+        wolf_gain_from_food=20,
+        grass=False,
+        grass_regrowth_time=30,
+        sheep_gain_from_food=4,
+        seed=None,
+    ):
+        """
+        Create a new Wolf-Sheep model with the given parameters.
+
+        Args:
+            initial_sheep: Number of sheep to start with
+            initial_wolves: Number of wolves to start with
+            sheep_reproduce: Probability of each sheep reproducing each step
+            wolf_reproduce: Probability of each wolf reproducing each step
+            wolf_gain_from_food: Energy a wolf gains from eating a sheep
+            grass: Whether to have the sheep eat grass for energy
+            grass_regrowth_time: How long it takes for a grass patch to regrow
+                                 once it is eaten
+            sheep_gain_from_food: Energy sheep gain from grass, if enabled.
+        """
+        super().__init__(seed=seed)
+        # Set parameters
+        self.width = width
+        self.height = height
+        self.initial_sheep = initial_sheep
+        self.initial_wolves = initial_wolves
+        self.grass = grass
+        self.grass_regrowth_time = grass_regrowth_time
+
+        self.grid = OrthogonalMooreGrid((self.width, self.height), torus=True)
+
+        collectors = {
+            "Wolves": lambda m: len(m.agents_by_type[Wolf]),
+            "Sheep": lambda m: len(m.agents_by_type[Sheep]),
+            "Grass": lambda m: len(m.agents_by_type[GrassPatch].select(lambda a:a.fully_grown)) if m.grass else -1,
+        }
+
+        self.datacollector = mesa.DataCollector(collectors)
+
+        # Create sheep:
+        for i in range(self.initial_sheep):
+            x = self.random.randrange(self.width)
```

Comment:
Don’t we have a nice random cell function for this?

(can also be done in another PR, but now that we’re looking at it)

Reply:
yes you can do `self.grid.all_cells.select_random_empty_cell()` or something close to that. 

I'll leave that for a seperate PR that goes into the code itself and cleans it up. Let's keep this focussed on the reorganization.